Locale awareness in Adaptive-Expressions

cosmicshuai opened the pull request:
This PR contains the work in #4025. Since some force-pushed commits in previous PR introduced too many stuffs in the timeline, I created this new PR and closed that one.
@tomlm @chrimc62 @cleemullins @boydc2014 @Danieladu 

1. Locale is obtained from 'turn.activity.locale' when an incoming activity is recieved and the locale info will be set to 'turn.locale' in memory. Also user can use TurnContext.Locale to get and set locale.

2. After a locale is set in TurnContext. It will be passed to EvaluationOptions in TemplateEngineLanguageGenerator. Then Options in Adaptive-Expressions will be updated. 

3. In locale related pre-built functions, the priority of locale follows: locale in APIs > locale from TurnContext > Thread.CurrentThread.CurrentCulture

Functions support locale paramter:
string(Object input, string locale?)
formatNumber(number input, integer precision, string locale?)
addToTime(string timestamp, integer interval, string unit, string format?, string locale?)
startOfDay(string timestamp, string format?, string locale?)
startOfHour(string timestamp, string format?, string locale?)
startOfMonth(string timestamp, string format?, string locale?)
convertToUTC(string timestamp, string timezone, string format?, string locale?)
convertFromUTC(string timestamp, string timezone, string format?, string locale?)
utcNow(string format?, string locale?)
getPastTime(integer interval, string unit, string format?, string locale?)
getFutureTime(integer interval, string unit, string format?, string locale?)
subtractFromTime(string timestamp, integer interval, string unit, string format?, string locale?)
formatEpoch(long epoch, string format?, string locale?)
formatTicks(long tictks, string format?, string locale?)
formatDateTime(string timestamp, string format?, string locale?)
addDays(string timestamp, int interval, string format?, string locale?)
addHours(string timestamp, int interval, string format?, string locale?)
addMinutes(string timestamp, int interval, string format?, string locale?)
toUpper(string input, string locale?)
toLower(string input, string locale?)
sentenceCase(string input, string locale?)
titleCase(string input, string locale?)

For all of these functions, locale parameter is optional. An locale paramter should be a valid locale string, such as "en-US", "en-GB", "fr-FR", "de-DE", "zh-CN", "ru-RU" etc.

chrimc62 commented on libraries/AdaptiveExpressions/Expression.cs:
```diff
@@ -440,7 +439,9 @@ public IReadOnlyList<string> References()
 
                     // filter references found in children 2 with iterator name
+#pragma warning disable CA1309 // Use ordinal stringcomparison
                     var nonLocalRefs2 = refs2.Where(x => !(x.Equals(iteratorName) || x.StartsWith(iteratorName + '.') || x.StartsWith(iteratorName + '[')))
```
Why not just use ordinal string comparison?

cosmicshuai replied:
fixed after merged master

chrimc62 commented on libraries/AdaptiveExpressions/Options.cs:
```diff
@@ -24,5 +27,13 @@ public Options(Options opt)
+        /// <value>
+        /// The locale info.
+        /// </value>
+        public string Locale { get; set; } = Thread.CurrentThread.CurrentCulture.Name;
```
>= Thread.CurrentThread.CurrentCulture.Name; [](start = 43, length = 43)

Won't this always be overridden by the constructor?  Either one or the other--I'd prefer the constructor.

cosmicshuai replied:
put the initilization in constructor. Removed it in gets and sets

chrimc62 commented on libraries/AdaptiveExpressions/BuiltinFunctions/ConvertFromUtc.cs:
```diff
@@ -23,24 +25,30 @@ private static (object value, string error) Evaluator(Expression expression, IMe
             string error = null;
             IReadOnlyList<object> args;
+            var locale = options.Locale != null ? new CultureInfo(options.Locale) : Thread.CurrentThread.CurrentCulture;
+            var format = FunctionUtils.DefaultDateTimeFormat;
```
Doesn't options.locale always have a locale of CurrentCulture by construction?  It would only be null if someone explicitly set it to null.

chrimc62 commented on libraries/AdaptiveExpressions/BuiltinFunctions/ConvertToUtc.cs:
```diff
@@ -23,24 +25,31 @@ private static (object value, string error) Evaluator(Expression expression, IMe
             string error = null;
             IReadOnlyList<object> args;
+            var locale = options.Locale != null ? new CultureInfo(options.Locale) : Thread.CurrentThread.CurrentCulture;
+            var format = FunctionUtils.DefaultDateTimeFormat;
```
Same issue with options.locale always having a value.

chrimc62 commented on libraries/AdaptiveExpressions/BuiltinFunctions/String.cs:
```diff
@@ -11,13 +14,48 @@ namespace AdaptiveExpressions.BuiltinFunctions
+                        {
+                            error = $"the second argument {args[1]} should be a locale string.";
+                        }
+                        else
```
Isn't this covered by DetermineLocale?

cosmicshuai replied:
move it to DetermineLocale

chrimc62 commented:
I generally like the changes.  Is there a work item to update all of the adaptive expression documentation for the new parameters?

chrimc62 commented on libraries/AdaptiveExpressions/FunctionUtils.cs:
```diff
@@ -1034,6 +1065,57 @@ internal static (object, string) NormalizeToDateTime(object timestamp, Func<Date
+                if (args.Count == maxArgsLength)
+                {
+                    (locale, error) = TryParseLocale(args[maxArgsLength - 1] as string);
+                }
```
Should verify is a string and otherwise have an error with the value.  Otherwise this will end up as null.

cosmicshuai replied:
add verify string, if not update error message

boydc2014 commented:
> I generally like the changes. Is there a work item to update all of the adaptive expression documentation for the new parameters?

The workflow is for us, developers make sure we have everything for doc in code as comments, and Emily will update the docs in bots-doc repo, @cosmicshuai you need to explicit about the parameters in comments.

cosmicshuai commented:
@chrimc62 @boydc2014 
I updated the PR description, which all the related functions' signature has been included. And the meaning of locale is also added.

gabog commented:
General comment on this feature, why don't we use the built in CultureInfo object rather than a string locale parameter? It will work seamlessly with any native string manipulation methods.

gabog commented on libraries/AdaptiveExpressions/BuiltinFunctions/ConvertFromUtc.cs:
```diff
@@ -23,24 +25,30 @@ private static (object value, string error) Evaluator(Expression expression, IMe
             object value = null;
             string error = null;
             IReadOnlyList<object> args;
+            var locale = options.Locale != null ? new CultureInfo(options.Locale) : Thread.CurrentThread.CurrentCulture;
```
Should we default to CurrentCulture or CurrentUICulture? See [this article](https://stackoverflow.com/questions/329033/what-is-the-difference-between-currentculture-and-currentuiculture-properties-of ) for additional details:

![image](https://user-images.githubusercontent.com/12264946/87802518-5abb7f00-c81f-11ea-9398-88a65b8f5e59.png)

cosmicshuai replied:
One of my concern this code is running on the server side. No matter CurrentCulture or CurrentUICulture was obtained from the server rather than reflecting the user's locale. 
If we have the user's culture info, it can be passed by the locale in incoming activities. The CurrentCulture used here is just a default value only if we could not determine user's locale.

gabog commented on libraries/AdaptiveExpressions/BuiltinFunctions/ConvertFromUtc.cs:
```diff
@@ -23,24 +25,30 @@ private static (object value, string error) Evaluator(Expression expression, IMe
             object value = null;
             string error = null;
             IReadOnlyList<object> args;
+            var locale = options.Locale != null ? new CultureInfo(options.Locale) : Thread.CurrentThread.CurrentCulture;
```
nit, in general I would call the variable and parameter cultureInfo rather than locale so it doesn't get confused with the locale string we use in the other places.

cosmicshuai replied:
Agree. One reason I choose locale is I want to use same property name in both C# and JS. 
Will discuss with Dong to choose a more accurate pamameter name here.

chrimc62 commented on libraries/AdaptiveExpressions/FunctionUtils.cs:
```diff
@@ -1062,6 +1095,67 @@ internal static (object, string) NormalizeToDateTime(object timestamp, Func<Date
+                    } 
+                    else
+                    {
+                        error = $"the {args[maxArgsLength - 1]} should be a locale string.";
```
>the  [](start = 34, length = 4)

Delete word.

chrimc62 commented:
> #pragma warning restore CA1801 // Review unused parameters

I think we should just break compatibility.  I doubt anyone is actually using these functions in custom functions and if they are they are a sophisticated developer.

---
Refers to: libraries/AdaptiveExpressions/FunctionUtils.cs:214 in 7d33aac. [](commit_id = 7d33aacfee34e3b93e0b80c10868fcaaffe2f22f, deletion_comment = False)

chrimc62 commented:
>                 throw new Exception($"foreach expect 3 parameters, found {expression.Children.Length}");

expects

---
Refers to: libraries/AdaptiveExpressions/FunctionUtils.cs:1047 in 7d33aac. [](commit_id = 7d33aacfee34e3b93e0b80c10868fcaaffe2f22f, deletion_comment = False)

chrimc62 commented on libraries/AdaptiveExpressions/FunctionUtils.cs:
```diff
@@ -1090,13 +1184,13 @@ internal static (object, string) ConvertTimeZoneFormat(string timezone)
+                result = datetime.ToString(format, locale);
             }
 #pragma warning disable CA1031 // Do not catch general exception types (capture any exception and return generic message)
             catch
```
Do you want to disable this for the whole file?

gabog replied:
I don't think we should, each exception should have a justification, in many cases we don't seem to put enough thought on what exceptions we are trying to catch and we are actually handling them.

cosmicshuai replied:
I think we should disable this for the whole file. Is there any recommended way to do that?

chrimc62 replied:
I'm fine with leaving per-instance.  I understand there is going to be a later clean-up looking at them all.

---
In reply to: [457435215](https://github.com/microsoft/botbuilder-dotnet/pull/4276#discussion_r457435215) [](ancestors = 457435215)

chrimc62 commented:
Just a few minor doc issues.

---
In reply to: [450832590](https://github.com/microsoft/botbuilder-dotnet/pull/4276#issuecomment-450832590) [](ancestors = 450832590)

gabog commented:
> > ```
> >             throw new Exception($"foreach expect 3 parameters, found {expression.Children.Length}");
> > ```
> 
> expects
> 
> Refers to: libraries/AdaptiveExpressions/FunctionUtils.cs:1047 in [7d33aac](https://github.com/microsoft/botbuilder-dotnet/commit/7d33aacfee34e3b93e0b80c10868fcaaffe2f22f). [](commit_id = [7d33aac](https://github.com/microsoft/botbuilder-dotnet/commit/7d33aacfee34e3b93e0b80c10868fcaaffe2f22f), deletion_comment = False)

One more comment, I haven't enabled this rule yet but we should not be throwing general Exceptions (see https://docs.microsoft.com/en-us/visualstudio/code-quality/ca2201?view=vs-2019 for more details).

chrimc62 commented:
In this case though adaptive expressions has worked very hard to not surface exceptions except through the error mechanism.  In the few places where we do catch exceptions, it is usually all exceptions for that reason.

From: Gabo Gilabert <notifications@github.com>
Sent: Friday, July 17, 2020 10:26 AM
To: microsoft/botbuilder-dotnet <botbuilder-dotnet@noreply.github.com>
Cc: Chris McConnell <chrimc@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [microsoft/botbuilder-dotnet] Locale awareness in Adaptive-Expressions (#4276)


@gabog commented on this pull request.

________________________________

In libraries/AdaptiveExpressions/FunctionUtils.cs<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fbotbuilder-dotnet%2Fpull%2F4276%23discussion_r456575634&data=02%7C01%7Cchrimc%40microsoft.com%7Ca85ef6eb06774481d38308d82a766f58%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637306035446976322&sdata=L11caSORAhOSJczfI8q9EpBsciMTkTy9dnDOHbpzqKM%3D&reserved=0>:

>          {

             string result = null;

             string error = null;

             try

             {

-                result = datetime.ToString(format, CultureInfo.InvariantCulture.DateTimeFormat);

+                result = datetime.ToString(format, locale);

             }

 #pragma warning disable CA1031 // Do not catch general exception types (capture any exception and return generic message)

             catch

I don't think we should, each exception should have a justification, in many cases we don't seem to put enough thought on what exceptions we are trying to catch and we are actually handling them.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmicrosoft%2Fbotbuilder-dotnet%2Fpull%2F4276%23discussion_r456575634&data=02%7C01%7Cchrimc%40microsoft.com%7Ca85ef6eb06774481d38308d82a766f58%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637306035446986315&sdata=dG%2BxrdRsXNZldaNj0bCC%2Fz6TuiE5fVa%2Bk0YRHKBlSc4%3D&reserved=0>, or unsubscribe<https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAEGSCMLLICPTYJXILKCFY23R4CCRPANCNFSM4O4DK2WA&data=02%7C01%7Cchrimc%40microsoft.com%7Ca85ef6eb06774481d38308d82a766f58%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637306035446986315&sdata=%2BChq9Ct28%2Bv3Yycslp4Eyy4yOMx15012WDfGyzgBdxM%3D&reserved=0>.

gabog commented:
Not sure about that @chrimc62 , take this snippet for example:

![image](https://user-images.githubusercontent.com/12264946/87815743-2605f280-c834-11ea-8fc2-f3898babb2ac.png)

We are "assuming" the exception was for an invalid locale, but there's other code in the block that could have thrown the exception (e.g.: `args[2] as string` could return null and CultureInfo would return InvalidCastException), the error message uses args[3] in the message (which I don't think is valid and it will throw an argument out of range).

There are other places where we blindly catch any exception, we return a generic message and we don't log the exceptions (we probably should for troubleshooting).

cosmicshuai commented:
> , but there's other

@gabog 
For type checking of arguments, we have both static checks in validator and dynamic checks during the runtime. 
There were some conversion errors(like Convert.ToInt32 method takes a long integer) in try catch blocks but did not throw the correct error. We have fixed these issues. 
In Adaptive-Expressions, we have to keep this way of catching errors so that we can provide more context of errors to users. 
Also in the next cycle, we will do a more thorough analysis on the error catching to aviod incorrect error messages.

mrivera-ms commented:
@carlosscastro, could you please take a look at this issue?

chrimc62 commented:
I was actually suggesting it should be included each functions class description--not the PR description.

---
In reply to: [659964303](https://github.com/microsoft/botbuilder-dotnet/pull/4276#issuecomment-659964303) [](ancestors = 659964303)

cosmicshuai commented:
> I was actually suggesting it should be included each functions class description--not the PR description.
> 
> In reply to: [659964303](https://github.com/microsoft/botbuilder-dotnet/pull/4276#issuecomment-659964303) [](ancestors = 659964303)

I will update the doc in each functions class description.

EricDahlvang commented:
@carlosscastro  can you take a look at this?

chrimc62 commented on libraries/AdaptiveExpressions/BuiltinFunctions/AddDays.cs:
```diff
@@ -7,6 +7,7 @@ namespace AdaptiveExpressions.BuiltinFunctions
 {
     /// <summary>
     /// Add a number of days to a timestamp.
+    /// AddDays function takes a timestamp string, an interval integer, an optional format string and an optional locale string.
```
>an optional format string and an optional locale string [](start = 72, length = 55)

Would be good to add what the optional parameters default too.

cosmicshuai replied:
I added some info about the default value for format and locale optional parameters.
Also removed the unnecessary see cref tag in EvaluationOptions class.

chrimc62 commented:
>     /// Options for evaluation of LG template <see cref="EvaluationOptions"/> class.

Remove this--it is referring to itself.

---
Refers to: libraries/Microsoft.Bot.Builder.LanguageGeneration/EvaluationOptions.cs:28 in ba2d253. [](commit_id = ba2d25352812be5cb6dc24db94f6c9822679f00e, deletion_comment = False)

boydc2014 commented:
Looks good to me and @chrimc62 now, can @mrivera-ms or @gabog help take a second look on this?

boydc2014 commented:
@carlosscastro can you help review this?

cosmicshuai commented:
@gabog @johnataylor 
Could you help review this PR, since it touched the code in bf-dialogs. Need your approval before it can be merged.

gabog commented on libraries/AdaptiveExpressions/FunctionUtils.cs:
```diff
@@ -585,6 +585,39 @@ public static EvaluateExpressionDelegate ApplyWithError(Func<IReadOnlyList<objec
+                    {
+                        (value, error) = function(args, options);
+                    }
+#pragma warning disable CA1031 // Do not catch general exception types
```
Always add a justification for a #pragma disable. Leave the rule description and add your justification in parenthesis so anyone that comes after you knows why you did it).

Example:
#pragma warning disable CA1031 // Do not catch general exception types (capture any exception and return it in the error)

cosmicshuai replied:
justification added

gabog commented on libraries/AdaptiveExpressions/FunctionUtils.cs:
```diff
@@ -1021,7 +1054,7 @@ internal static void ValidateForeach(Expression expression)
             {
-                throw new Exception($"foreach expect 3 parameters, found {expression.Children.Length}");
+                throw new Exception($"foreach expects 3 parameters, found {expression.Children.Length}");
             }
```
Do not throw general exceptions, in this cause, ArgumentException would be a better thing to throw.

cosmicshuai replied:
changed to ArgumentException

gabog commented on libraries/AdaptiveExpressions/FunctionUtils.cs:
```diff
@@ -1072,6 +1105,67 @@ internal static (object, string) NormalizeToDateTime(object timestamp, Func<Date
+            {
+                result = new CultureInfo(localeStr);
+            }
+#pragma warning disable CA1031 // Do not catch general exception types
```
Add justification

cosmicshuai replied:
change to the CultureNotFoundException. Also changed the naming of localeStr to locale

gabog commented on libraries/Microsoft.Bot.Builder/BotAdapter.cs:
```diff
@@ -224,6 +225,22 @@ protected async Task RunPipelineAsync(ITurnContext turnContext, BotCallbackHandl
+                        (turnContext as TurnContext).Locale = turnContext.Activity.Locale;
+                    }
+#pragma warning disable CA1031 // Do not catch general exception types
+                    catch
```
Refactor this, do not catch all exceptions, in this case it looks like you are trying to catch an exception that may occur if you send an invalid locale code and try to create a cultureInfo instance with that. Catch that specific exception instead and do not add the pragma.

cosmicshuai replied:
change to CultureNotFoundException

gabog commented on libraries/Microsoft.Bot.Builder/BotAdapter.cs:
```diff
@@ -224,6 +225,22 @@ protected async Task RunPipelineAsync(ITurnContext turnContext, BotCallbackHandl
+                    {
+                        // if turnContext.Activity.Locale is illegal, then TurnContext.Locale will set to en-US as default. 
+                        (turnContext as TurnContext).Locale = "en-US";
+                    }
```
Do not hardcode en-US, default to the servers or machine CurrentCulture instead

cosmicshuai replied:
change to the Thread.CurrentThread.CurrentCulture.Name

gabog commented on libraries/Microsoft.Bot.Builder/BotAdapter.cs:
```diff
@@ -224,6 +225,22 @@ protected async Task RunPipelineAsync(ITurnContext turnContext, BotCallbackHandl
+                if (turnContext.Activity.Locale != null)
+                {
+                    try
+                    {
```
Add unit tests for this new logic

cosmicshuai replied:
unit tests added in BotAdapterTests

gabog commented on libraries/Microsoft.Bot.Builder/TurnContext.cs:
```diff
@@ -51,6 +53,16 @@ public TurnContext(BotAdapter adapter, Activity activity)
         /// <value>The services registered on this context object.</value>
         public TurnContextStateCollection TurnState { get; } = new TurnContextStateCollection();
 
+        /// <summary>
```
Has this new property been validated by Tom, Steve, etc? In general we are using the activity locale in our code to drive localization, do we really need a new context property?

cosmicshuai replied:
Yes, this new property is based on Tom's opinion. A context property can help the locale to be passed to adaptive-expressions

gabog commented on libraries/Microsoft.Bot.Builder.Dialogs/DialogExtensions.cs:
```diff
@@ -95,6 +95,17 @@ public static async Task RunAsync(this Dialog dialog, ITurnContext turnContext,
+        /// <param name="dialogContext">The dialogContext to extract information.</param>
+        /// <returns>A string representing the current locale.</returns>
+        public static string GetLocale(this DialogContext dialogContext)
+        {
```
We don't need an extension method for this, this method can be added to DialogContext that is in the same assembly as this extension method. Move it there

cosmicshuai replied:
Moved it to DialogContext class

gabog commented on libraries/Microsoft.Bot.Builder.Dialogs/DialogExtensions.cs:
```diff
@@ -95,6 +95,17 @@ public static async Task RunAsync(this Dialog dialog, ITurnContext turnContext,
+        public static string GetLocale(this DialogContext dialogContext)
+        {
+            var locale = (dialogContext.Context as TurnContext).Locale ?? null;
+            return locale;
```
Avoid soft casts, it will return null if Context is not of the expected type and you will get a null ref exception.

I would refactor this line of code as:
return ((TurnContext)dialogContext.Context).Locale

But see my note about adding a new Locale property in TurnContext

If you decide to keep this method add unit tests for it.

boydc2014 replied:
I think this is followed by the previous comment Tom gave, and sure @cosmicshuai we should add UT for this.

cosmicshuai replied:
removed the soft cast here.

gabog replied:
Did you add the tests?

cosmicshuai replied:
I didn't find a appropriate place to add a unit test for this. Do you have any suggestion for this? @gabog 
But I have a integration test for this logic in Adaptive-Templates-Tests.

gabog replied:
The convention is to put unit tests un a file that has the same name of the class being tested + test. If we don't have, we should create one.
Assuming you moved the code into dialogcontext, the test should be in a file called DialogContextTests.cs

cosmicshuai replied:
Got it. I will create a new PR to add test for this.
